front/gatsby/gatsby-browser.js

```diff
@@ -0,0 +1,38 @@
+// Copyright 2018-2020 @paritytech/Nomidot authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import { ApolloProvider } from '@apollo/react-hooks';
+import { WsProvider } from '@polkadot/api';
+import {
+  AccountsContextProvider,
+  ApiContextProvider,
+} from '@substrate/context';
+import { global } from '@substrate/design-system';
+import React from 'react';
+
+import client from './src/apollo';
+import { Layout, Seo } from './src/components';
+import { APP_SLUG } from './src/util';
+
+const { GlobalStyle } = global;
+
+export const wrapRootElement = ({ element }) => (
+  <ApolloProvider client={client}>
+    <ApiContextProvider
+      provider={new WsProvider('wss://cc3-5.kusama.network/')}
+    >
+      <AccountsContextProvider originName={APP_SLUG}>
+        {element}
+      </AccountsContextProvider>
+    </ApiContextProvider>
+  </ApolloProvider>
+);
+
+export const wrapPageElement = ({ element }) => (
+  <Layout>
+    <Seo title='Polkadot/Kusama Staking Portal' />
+    <GlobalStyle />
+    {element}
+  </Layout>
```

Comment:
Having a `wrapRootElement` is still okay, because I think this rootElement will really be the same on all pages

However, `wrapPageElement`'s SEO might be different on each page, e.g. if we want a validator's name to appear in the title.

Reply:
A TS-only alternative would be:

```tsx
export function RootElement() {
  <AllTheContexts>
    <Layout, SEO, GlobalStyle />
  </AllTheContext>
}

// in pages/something.tsx
export function MyPage() {
  <RootElement seoTitle='MyPage'>
    {page-specific-UI}
  <RootElemetn>
}
```

Reply:
`wrapPageElement` can receive props data about the location (and others from RouteComponentProps) from `@reach/router` so we can later pass down the appropriate SEO based on that. From what I've read, `wrapRootElement` is the Gatsby idiomatic place to put most if not all Providers, and `wrapPageElement` the place to put wrapper components that won’t get unmounted on page changes.

1. perhaps we can work around it with: https://github.com/gatsbyjs/gatsby/issues/1457#issuecomment-381405638
2.  I don't think that it is in itself a reason to go the long winded non-Gatsby way, considering the selling point of Gatsby is its magical wonders.

Reply:
docs for reference: https://www.gatsbyjs.org/docs/browser-apis/#wrapPageElement